src/net/larse/lcms/algorithms/VCT.java

```diff
@@ -426,70 +340,49 @@ private void initializePixel() {
-    int pctGoodObs = 0;
     int badCount = 0;
+    
     // Flag each year's pixel as good or bad based on the mask value
     // TODO: Note that class 0 (BACKGROUND) is *not* being flagged as
     // "fillable" in the original source code (ie. SLC-off errors).
     // We need to ask Cheng about this
-
-    //YANG: refactory of the code logic to speed computation.
-    int tmp = this.mask[0];
-    if (tmp != 0 && tmp <= FILL_CLASSES) {
-      this.qFlag[0] = QA_BAD;
-      badCount++;
-    } else {
-      this.qFlag[0] = QA_GOOD;
-    }
-
-    tmp = this.mask[this.numYears - 1];
-    if (tmp != 0 && tmp <= FILL_CLASSES) {
-      this.qFlag[this.numYears - 1] = QA_BAD;
-      badCount++;
-    } else {
-      this.qFlag[this.numYears - 1] = QA_GOOD;
-    }
-
-    // Look for spikes and dips that may be unflagged cloud and shadow and
-    // set the QA flag accordingly
-    for (int i = 1; i < this.numYears - 1; i++) {
-      if (this.mask[i] != 0 && this.mask[i] <= FILL_CLASSES) {
+    //
+    // Also, look for spikes and dips that may be unflagged cloud and
+    // shadow and set the QA flag accordingly
+    for (int i = 0; i < this.numYears; i++) {
+      
+      // Start by calling the value good
+      this.qFlag[i] = QA_GOOD;
+      
+      // Check for bad mask value
+      if (mask[i] != 0 && mask[i] <= FILL_CLASSES) {
         this.qFlag[i] = QA_BAD;
         badCount++;
-
-        // Skip over pixels already labeled as QA_BAD
         continue;
-      } else {
-        this.qFlag[i] = QA_GOOD;
       }
-
-      // Pixels higher in UD than neighboring years - cloud
-      // Pixels lower than UD than neighboring years - shadow
-      if (isRelativeCloud(i) || isRelativeShadow(i)) {
+      
+      // Check for cloud/shadow
+      if (i == 0) {
+        if (isBadEndpoint(i, i + 1)) {
+          this.qFlag[i] = QA_BAD;
+          badCount++;
+        }
+      } else if (i == this.numYears - 1) {
+        if (isBadEndpoint(i, i - 1)) {
+          this.qFlag[i] = QA_BAD;
+          badCount++;
+        }
+      } else if (isRelativeCloud(i) || isRelativeShadow(i)) {
         this.qFlag[i] = QA_BAD;
         badCount++;
       }
     }
 
-    // Now test the start/end conditions
-    int n = this.numYears;
-    if (this.qFlag[0] == QA_GOOD && isBadEndpoint(0, 1)) {
-      this.qFlag[0] = QA_BAD;
-      badCount++;
-    }
-    if (this.qFlag[n - 1] == QA_GOOD && isBadEndpoint(n - 1, n - 2)) {
-      this.qFlag[n - 1] = QA_BAD;
-      badCount++;
-    }
-
-    // Calculate bad percentage
-    pctGoodObs = (int) (100.0 - ((100.0 * badCount) / this.numYears));
-
     // Interpolate for bad observations indicated by qFlag
-    //TODO: (yang) is there any special treatment when percentGood is less than 50%?
+    // TODO: (yang) is there any special treatment when percentGood is less
+    // than 50%?
     int i = 0;
-    if (pctGoodObs > 50.0) {
+    if (badCount <= (this.numYears / 2.0)) {
```

Comment:
Still need to check what happens when we have badCount that exceeds half the number of years